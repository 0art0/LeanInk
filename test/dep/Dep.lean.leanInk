[{"contents": "import mathlib\n\n#check ", "_type": "text"},
 {"messages":
  [{"contents": "Dep.lean:3:7: error: unknown identifier 'Nat'\n",
    "_type": "message"}],
  "goals": [],
  "contents": "",
  "_type": "sentence"},
 {"messages":
  [{"contents": "Dep.lean:3:7: error: unknown identifier 'Nat'\n",
    "_type": "message"},
   {"contents": "Dep.lean:3:0: error: unknown constant 'sorryAx'\n",
    "_type": "message"}],
  "goals": [],
  "contents": "#check Nat",
  "_type": "sentence"},
 {"messages":
  [{"contents": "Dep.lean:3:7: error: unknown identifier 'Nat'\n",
    "_type": "message"}],
  "goals": [],
  "contents": "",
  "_type": "sentence"},
 {"contents": "", "_type": "text"}]