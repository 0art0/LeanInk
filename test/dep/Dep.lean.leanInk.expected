[{"contents":
  "import Mathlib.Tactic.Basic\n\nexample : ‚àÄ a b : Nat, a = b ‚Üí b = a := ",
  "_type": "text"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses": [],
    "conclusion": "‚àÄ (a b : Nat), a = b ‚Üí b = a",
    "_type": "goal"}],
  "contents": "by\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["a", "b"], "body": "", "_type": "hypothesis"},
     {"type": "a = b", "names": ["h"], "body": "", "_type": "hypothesis"}],
    "conclusion": "b = a",
    "_type": "goal"}],
  "contents": "introv h",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses": [],
    "conclusion": "‚àÄ (a b : Nat), a = b ‚Üí b = a",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses": [],
    "conclusion": "Goals accomplished! üêô",
    "_type": "goal"}],
  "contents": "exact h.symm",
  "_type": "sentence"},
 {"contents": "\n\nexample (n : Nat) : n = n := ", "_type": "text"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n"], "body": "", "_type": "hypothesis"}],
    "conclusion": "n = n",
    "_type": "goal"}],
  "contents": "by\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "zero",
    "hypotheses": [],
    "conclusion": "Nat.zero = Nat.zero",
    "_type": "goal"},
   {"name": "succ",
    "hypotheses":
    [{"type": "Nat", "names": ["n‚úù"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Nat.succ n‚úù = Nat.succ n‚úù",
    "_type": "goal"}],
  "contents": "induction n",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n"], "body": "", "_type": "hypothesis"}],
    "conclusion": "n = n",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses": [],
    "conclusion": "Goals accomplished! üêô",
    "_type": "goal"}],
  "contents": "exacts [rfl, rfl]",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n"], "body": "", "_type": "hypothesis"}],
    "conclusion": "n = n",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses": [],
    "conclusion": "Goals accomplished! üêô",
    "_type": "goal"}],
  "contents": "exacts []",
  "_type": "sentence"},
 {"contents": "\n\nexample (n : Nat) : Nat := ", "_type": "text"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Nat",
    "_type": "goal"}],
  "contents": "by\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Nat",
    "_type": "goal"}],
  "contents": "guard_hyp n : Nat",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Nat",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n"], "body": "", "_type": "hypothesis"},
     {"type": "Nat", "names": ["m"], "body": "1", "_type": "hypothesis"}],
    "conclusion": "Nat",
    "_type": "goal"}],
  "contents": "let m : Nat := 1",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Nat",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n"], "body": "", "_type": "hypothesis"},
     {"type": "Nat", "names": ["m"], "body": "1", "_type": "hypothesis"}],
    "conclusion": "Nat",
    "_type": "goal"}],
  "contents": "guard_hyp m := 1",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Nat",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n"], "body": "", "_type": "hypothesis"},
     {"type": "Nat", "names": ["m"], "body": "1", "_type": "hypothesis"}],
    "conclusion": "Nat",
    "_type": "goal"}],
  "contents": "guard_hyp m : Nat := 1",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Nat",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n"], "body": "", "_type": "hypothesis"},
     {"type": "Nat", "names": ["m"], "body": "1", "_type": "hypothesis"}],
    "conclusion": "Nat",
    "_type": "goal"}],
  "contents": "guard_target == Nat",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Nat",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses": [],
    "conclusion": "Goals accomplished! üêô",
    "_type": "goal"}],
  "contents": "exact 0",
  "_type": "sentence"},
 {"contents": "\n\nexample (a b : Nat) : a ‚â† b ‚Üí ¬¨ a = b := ", "_type": "text"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["a", "b"], "body": "", "_type": "hypothesis"}],
    "conclusion": "a ‚â† b ‚Üí ¬¨a = b",
    "_type": "goal"}],
  "contents": "by\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["a", "b"], "body": "", "_type": "hypothesis"}],
    "conclusion": "¬¨a = b",
    "_type": "goal"}],
  "contents": "intros",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["a", "b"], "body": "", "_type": "hypothesis"}],
    "conclusion": "a ‚â† b ‚Üí ¬¨a = b",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["a", "b"], "body": "", "_type": "hypothesis"}],
    "conclusion": "¬¨a = b",
    "_type": "goal"}],
  "contents": "by_contra H",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["a", "b"], "body": "", "_type": "hypothesis"}],
    "conclusion": "a ‚â† b ‚Üí ¬¨a = b",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses": [],
    "conclusion": "Goals accomplished! üêô",
    "_type": "goal"}],
  "contents": "contradiction",
  "_type": "sentence"},
 {"contents": "\n\nexample (a b : Nat) : ¬¨¬¨ a = b ‚Üí a = b := ",
  "_type": "text"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["a", "b"], "body": "", "_type": "hypothesis"}],
    "conclusion": "¬¨¬¨a = b ‚Üí a = b",
    "_type": "goal"}],
  "contents": "by\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["a", "b"], "body": "", "_type": "hypothesis"}],
    "conclusion": "a = b",
    "_type": "goal"}],
  "contents": "intros",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["a", "b"], "body": "", "_type": "hypothesis"}],
    "conclusion": "¬¨¬¨a = b ‚Üí a = b",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "h",
    "hypotheses":
    [{"type": "Nat", "names": ["a", "b"], "body": "", "_type": "hypothesis"},
     {"type": "¬¨a = b", "names": ["H"], "body": "", "_type": "hypothesis"}],
    "conclusion": "False",
    "_type": "goal"}],
  "contents": "by_contra H",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["a", "b"], "body": "", "_type": "hypothesis"}],
    "conclusion": "¬¨¬¨a = b ‚Üí a = b",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses": [],
    "conclusion": "Goals accomplished! üêô",
    "_type": "goal"}],
  "contents": "contradiction",
  "_type": "sentence"},
 {"contents": "\n\nexample (p q : Prop) : ¬¨¬¨ p ‚Üí p := ", "_type": "text"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "¬¨¬¨p ‚Üí p",
    "_type": "goal"}],
  "contents": "by\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p",
    "_type": "goal"}],
  "contents": "intros",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "¬¨¬¨p ‚Üí p",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "h",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"},
     {"type": "¬¨p", "names": ["H"], "body": "", "_type": "hypothesis"}],
    "conclusion": "False",
    "_type": "goal"}],
  "contents": "by_contra H",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "¬¨¬¨p ‚Üí p",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses": [],
    "conclusion": "Goals accomplished! üêô",
    "_type": "goal"}],
  "contents": "contradiction",
  "_type": "sentence"},
 {"contents": "\n\n-- Test `iterate n ...`\nexample (n m : Nat) : Unit := ",
  "_type": "text"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n", "m"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Unit",
    "_type": "goal"}],
  "contents": "by\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "zero",
    "hypotheses":
    [{"type": "Nat", "names": ["m"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Unit",
    "_type": "goal"},
   {"name": "succ",
    "hypotheses":
    [{"type": "Nat", "names": ["m"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Unit",
    "_type": "goal"}],
  "contents": "cases n",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n", "m"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Unit",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "zero.zero",
    "hypotheses": [],
    "conclusion": "Unit",
    "_type": "goal"},
   {"name": "zero.succ",
    "hypotheses": [],
    "conclusion": "Unit",
    "_type": "goal"},
   {"name": "succ",
    "hypotheses":
    [{"type": "Nat", "names": ["m"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Unit",
    "_type": "goal"}],
  "contents": "cases m",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n", "m"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Unit",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses": [],
    "conclusion": "Goals accomplished! üêô",
    "_type": "goal"}],
  "contents": "iterate 3 exact ()",
  "_type": "sentence"},
 {"contents":
  "\n\n-- Test `iterate ...`, which should repeat until failure.\nexample (n m : Nat) : Unit := ",
  "_type": "text"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n", "m"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Unit",
    "_type": "goal"}],
  "contents": "by\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "zero",
    "hypotheses":
    [{"type": "Nat", "names": ["m"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Unit",
    "_type": "goal"},
   {"name": "succ",
    "hypotheses":
    [{"type": "Nat", "names": ["m"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Unit",
    "_type": "goal"}],
  "contents": "cases n",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n", "m"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Unit",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "zero.zero",
    "hypotheses": [],
    "conclusion": "Unit",
    "_type": "goal"},
   {"name": "zero.succ",
    "hypotheses": [],
    "conclusion": "Unit",
    "_type": "goal"},
   {"name": "succ",
    "hypotheses":
    [{"type": "Nat", "names": ["m"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Unit",
    "_type": "goal"}],
  "contents": "cases m",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n", "m"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Unit",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses": [],
    "conclusion": "Goals accomplished! üêô",
    "_type": "goal"}],
  "contents": "iterate exact ()",
  "_type": "sentence"},
 {"contents": "\n\nexample (n : Nat) : Nat := ", "_type": "text"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Nat",
    "_type": "goal"}],
  "contents": "by\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Nat",
    "_type": "goal"}],
  "contents": "iterate exact ()",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Nat", "names": ["n"], "body": "", "_type": "hypothesis"}],
    "conclusion": "Nat",
    "_type": "goal"}],
  "contents": " -- silently succeeds, after iterating 0 times\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses": [],
    "conclusion": "Goals accomplished! üêô",
    "_type": "goal"}],
  "contents": "iterate exact n",
  "_type": "sentence"},
 {"contents":
  "\n\nexample (p q r s : Prop) : p ‚Üí q ‚Üí r ‚Üí s ‚Üí (p ‚àß q) ‚àß (r ‚àß s ‚àß p) ‚àß (p ‚àß r ‚àß q) := ",
  "_type": "text"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Prop",
      "names": ["p", "q", "r", "s"],
      "body": "",
      "_type": "hypothesis"}],
    "conclusion": "p ‚Üí q ‚Üí r ‚Üí s ‚Üí (p ‚àß q) ‚àß (r ‚àß s ‚àß p) ‚àß p ‚àß r ‚àß q",
    "_type": "goal"}],
  "contents": "by\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Prop",
      "names": ["p", "q", "r", "s"],
      "body": "",
      "_type": "hypothesis"}],
    "conclusion": "(p ‚àß q) ‚àß (r ‚àß s ‚àß p) ‚àß p ‚àß r ‚àß q",
    "_type": "goal"}],
  "contents": "intros",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Prop",
      "names": ["p", "q", "r", "s"],
      "body": "",
      "_type": "hypothesis"}],
    "conclusion": "p ‚Üí q ‚Üí r ‚Üí s ‚Üí (p ‚àß q) ‚àß (r ‚àß s ‚àß p) ‚àß p ‚àß r ‚àß q",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Prop",
      "names": ["p", "q", "r", "s"],
      "body": "",
      "_type": "hypothesis"}],
    "conclusion": "(p ‚àß q) ‚àß (r ‚àß s ‚àß p) ‚àß p ‚àß r ‚àß q",
    "_type": "goal"}],
  "contents": "repeat' ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "right.left",
    "hypotheses":
    [{"type": "Prop",
      "names": ["p", "q", "r", "s"],
      "body": "",
      "_type": "hypothesis"}],
    "conclusion": "r ‚àß s ‚àß p",
    "_type": "goal"},
   {"name": "right.right",
    "hypotheses":
    [{"type": "Prop",
      "names": ["p", "q", "r", "s"],
      "body": "",
      "_type": "hypothesis"}],
    "conclusion": "p ‚àß r ‚àß q",
    "_type": "goal"}],
  "contents": "constructor",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Prop",
      "names": ["p", "q", "r", "s"],
      "body": "",
      "_type": "hypothesis"}],
    "conclusion": "p ‚Üí q ‚Üí r ‚Üí s ‚Üí (p ‚àß q) ‚àß (r ‚àß s ‚àß p) ‚àß p ‚àß r ‚àß q",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "left.left",
    "hypotheses":
    [{"type": "Prop",
      "names": ["p", "q", "r", "s"],
      "body": "",
      "_type": "hypothesis"}],
    "conclusion": "p",
    "_type": "goal"},
   {"name": "left.right",
    "hypotheses":
    [{"type": "Prop",
      "names": ["p", "q", "r", "s"],
      "body": "",
      "_type": "hypothesis"}],
    "conclusion": "q",
    "_type": "goal"},
   {"name": "right.left.left",
    "hypotheses":
    [{"type": "Prop",
      "names": ["p", "q", "r", "s"],
      "body": "",
      "_type": "hypothesis"}],
    "conclusion": "r",
    "_type": "goal"},
   {"name": "right.left.right.left",
    "hypotheses":
    [{"type": "Prop",
      "names": ["p", "q", "r", "s"],
      "body": "",
      "_type": "hypothesis"}],
    "conclusion": "s",
    "_type": "goal"},
   {"name": "right.left.right.right",
    "hypotheses":
    [{"type": "Prop",
      "names": ["p", "q", "r", "s"],
      "body": "",
      "_type": "hypothesis"}],
    "conclusion": "p",
    "_type": "goal"},
   {"name": "right.right.left",
    "hypotheses":
    [{"type": "Prop",
      "names": ["p", "q", "r", "s"],
      "body": "",
      "_type": "hypothesis"}],
    "conclusion": "p",
    "_type": "goal"},
   {"name": "right.right.right.left",
    "hypotheses":
    [{"type": "Prop",
      "names": ["p", "q", "r", "s"],
      "body": "",
      "_type": "hypothesis"}],
    "conclusion": "r",
    "_type": "goal"},
   {"name": "right.right.right.right",
    "hypotheses":
    [{"type": "Prop",
      "names": ["p", "q", "r", "s"],
      "body": "",
      "_type": "hypothesis"}],
    "conclusion": "q",
    "_type": "goal"}],
  "contents": "repeat' ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses": [],
    "conclusion": "Goals accomplished! üêô",
    "_type": "goal"}],
  "contents": "assumption",
  "_type": "sentence"},
 {"contents": "\n\nexample (p q : Prop) : p ‚Üí q ‚Üí (p ‚àß q) ‚àß (p ‚àß q ‚àß p) := ",
  "_type": "text"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p ‚Üí q ‚Üí (p ‚àß q) ‚àß p ‚àß q ‚àß p",
    "_type": "goal"}],
  "contents": "by\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "(p ‚àß q) ‚àß p ‚àß q ‚àß p",
    "_type": "goal"}],
  "contents": "intros",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p ‚Üí q ‚Üí (p ‚àß q) ‚àß p ‚àß q ‚àß p",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "left",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p ‚àß q",
    "_type": "goal"},
   {"name": "right",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p ‚àß q ‚àß p",
    "_type": "goal"}],
  "contents": "constructor",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p ‚Üí q ‚Üí (p ‚àß q) ‚àß p ‚àß q ‚àß p",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "left",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p ‚àß q",
    "_type": "goal"},
   {"name": "right",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p ‚àß q ‚àß p",
    "_type": "goal"}],
  "contents": "fail_if_success ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "left",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p ‚àß q",
    "_type": "goal"},
   {"name": "right",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p ‚àß q ‚àß p",
    "_type": "goal"}],
  "contents": "any_goals ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "right",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p ‚àß q ‚àß p",
    "_type": "goal"}],
  "contents": "assumption",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p ‚Üí q ‚Üí (p ‚àß q) ‚àß p ‚àß q ‚àß p",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "left",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p ‚àß q",
    "_type": "goal"},
   {"name": "right",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p ‚àß q ‚àß p",
    "_type": "goal"}],
  "contents": "all_goals ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "left.left",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p",
    "_type": "goal"},
   {"name": "left.right",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "q",
    "_type": "goal"}],
  "contents": "constructor",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p ‚Üí q ‚Üí (p ‚àß q) ‚àß p ‚àß q ‚àß p",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "left.left",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p",
    "_type": "goal"},
   {"name": "left.right",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "q",
    "_type": "goal"},
   {"name": "right.left",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p",
    "_type": "goal"},
   {"name": "right.right",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "q ‚àß p",
    "_type": "goal"}],
  "contents": "any_goals ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "right.right",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "q ‚àß p",
    "_type": "goal"}],
  "contents": "assumption",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p ‚Üí q ‚Üí (p ‚àß q) ‚àß p ‚àß q ‚àß p",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "right.right.left",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "q",
    "_type": "goal"},
   {"name": "right.right.right",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p",
    "_type": "goal"}],
  "contents": "constructor",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p ‚Üí q ‚Üí (p ‚àß q) ‚àß p ‚àß q ‚àß p",
    "_type": "goal"}],
  "contents": "\n  ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "right.right.left",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "q",
    "_type": "goal"},
   {"name": "right.right.right",
    "hypotheses":
    [{"type": "Prop", "names": ["p", "q"], "body": "", "_type": "hypothesis"}],
    "conclusion": "p",
    "_type": "goal"}],
  "contents": "any_goals ",
  "_type": "sentence"},
 {"messages": [],
  "goals":
  [{"name": "",
    "hypotheses": [],
    "conclusion": "Goals accomplished! üêô",
    "_type": "goal"}],
  "contents": "assumption",
  "_type": "sentence"},
 {"contents": "", "_type": "text"}]